SignalServiceKit/src/Loki/API/LokiAPI.swift

```diff
@@ -48,36 +50,42 @@ import PromiseKit
         }.retryingIfNeeded(maxRetryCount: maxRetryCount).map { Set($0) }
     }
     
-    public static func sendSignalMessage(_ signalMessage: SignalMessage, to destination: String, timestamp: UInt64) -> Promise<Set<Promise<RawResponse>>> {
-        let isP2PMessagingPossible = false
-        return Message.from(signalMessage: signalMessage, timestamp: timestamp, requiringPoW: !isP2PMessagingPossible).then(sendMessage)
-    }
-    
-    public static func sendMessage(_ lokiMessage: Message) -> Promise<Set<Promise<RawResponse>>> {
-        let isP2PMessagingPossible = false
-        if isP2PMessagingPossible {
-            // TODO: Send using P2P protocol
-        } else {
+    public static func sendSignalMessage(_ signalMessage: SignalMessage, to destination: String, with timestamp: UInt64) -> Promise<Set<Promise<RawResponse>>> {
+        guard let lokiMessage = Message.from(signalMessage: signalMessage, timestamp: timestamp) else { return Promise(error: Error.messageConversionFailed) }
+        let destination = lokiMessage.destination
+        func sendLokiMessage(_ lokiMessage: Message, to targets: [Target]) -> Promise<Set<Promise<RawResponse>>> {
             let parameters = lokiMessage.toJSON()
-            return getTargetSnodes(for: lokiMessage.destination).mapValues { invoke(.sendMessage, on: $0, associatedWith: lokiMessage.destination, parameters: parameters) }
-                .retryingIfNeeded(maxRetryCount: maxRetryCount).map { Set($0) }
+            return Promise.value(targets).mapValues { invoke(.sendMessage, on: $0, associatedWith: destination, parameters: parameters) }.map { Set($0) }
         }
-    }
-    
-    public static func ping(_ hexEncodedPublicKey: String) -> Promise<Set<Promise<RawResponse>>> {
-        let isP2PMessagingPossible = false
-        if isP2PMessagingPossible {
-            // TODO: Send using P2P protocol
+        func sendLokiMessageUsingSwarmAPI() -> Promise<Set<Promise<RawResponse>>> {
+            return lokiMessage.calculatePoW().then { updatedLokiMessage -> Promise<Set<Promise<RawResponse>>> in
```

Comment:
We can actually speed this up a bunch by doing the PoW calculation while getting the swarm